### PR TITLE
NATIVE-556: Build configs for xcode

### PR DESCRIPTION
**Pull request by klinzo**
This pull request belongs to an issue on our [bugtracker](https://issues.integreat-app.de/).
You can find it there by looking for an issue with the key which is mentioned in the title of this pull request.
It starts with the keyword **NATIVE**.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -525,8 +562,8 @@ jobs:
...
       - run:
-          name: "[FL] E2E Tests on Browserstack"
-          command: bundle exec fastlane android browserstack_e2e_tests
+          name: "[FL] Android E2E Tests on Browserstack"
+          command: bundle exec fastlane android browserstack_e2e_tests app_name:integreat-test-cms
       - run:
-          name: "[FL] E2E Tests on Browserstack"
-          command: bundle exec fastlane ios browserstack_e2e_tests
+          name: "[FL] iOS E2E Tests on Browserstack"
+          command: bundle exec fastlane ios browserstack_e2e_tests app_name:integreat-test-cms
```
Comment (Schedulaar) [inline .circleci/config.yml:569]: Can we split this file into multiple? This file is too large.

Comment (klinzo) [reply .circleci/config.yml:569]: I'll look into yet. I think it is possible.

Comment (klinzo) [reply .circleci/config.yml:569]: Created an issue :)

File: build-configs/configs/integreat-test-cms/index.js [hunk 2]
```diff
@@ -6,7 +6,10 @@ import type { BuildConfigType } from '../BuildConfigType'
 const IntegreatTestCmsBuildConfig: BuildConfigType = {
   ...IntegreatBuildConfig,
   cmsUrl: 'https://cms-test.integreat-app.de',
-  switchCmsUrl: 'https://cms.integreat-app.de'
+  switchCmsUrl: 'https://cms.integreat-app.de',
+  iosBuildOptions: {
```
Comment (klinzo) [inline build-configs/configs/integreat-test-cms/index.js:10]: Not sure about naming here. Any ideas?

Comment (Schedulaar) [reply build-configs/configs/integreat-test-cms/index.js:10]: I don't see how this is useful. Why is `BUILD_CONFIG_APP_NAME` different to `integreat-test-cms` (build-config-id) AND `Integreat` (appName)? What kind of app name is it then?

If it's because we want to show the dev different names, I'd prefer simply switching the `appName` in the `integreat-test-cms` BuildConfig to `IntegreatTestCms`.

File: fastlane/Fastfile [hunk 3]
```diff
@@ -74,10 +71,11 @@ platform :android do
   lane :sentry_upload do |options|
     version_code = options[:version_code]
     version_name = options[:version_name]
+    app_name = options[:app_name]
 
     generic_sentry_upload(
-        package_name: 'tuerantuer.app.integreat',
-        sourcemap_path: './index.android.bundle.map',
+        package_name: "tuerantuer.app.integreat",
```
[the remark this answers is not in the document]
Comment (klinzo) [reply fastlane/Fastfile:77]: Will be done in a separate task. This task is just about making build configs usable.

File: tools/create-xcconfigs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+const createXcconfigs = async buildConfigName => {
+  const { iosBuildOptions } = loadBuildConfig(buildConfigName)
+  const content = Object.keys(iosBuildOptions).map(key => `${key} = ${iosBuildOptions[key]}`).join('\n')
+  fs.writeFileSync('ios/buildConfig.tmp.xcconfig', content)
+}
+
+program
+  .command('config <build_config_name>')
+  .description('creates and writes')
```
[the remark this answers is not in the document]
Comment (klinzo) [reply tools/create-xcconfigs:18]: Uhm yeah :D

File: build-configs/configs/integreat-test-cms/index.js (hunk 2 above)
Comment (maxammann) [reply build-configs/configs/integreat-test-cms/index.js:10]: I'd go with only `ios` or `native-ios`. It is already clear that we deal with build options as `switchCmsUrl` is also a Build Option, right?

File: build-configs/configs/BuildConfigType.js [hunk 5]
```diff
@@ -10,5 +10,8 @@ export type BuildConfigType = {|
   allowedHostNames: Array<string>,
   featureFlags: FeatureFlagsType,
   lightTheme: ThemeType,
-  darkTheme: ThemeType
+  darkTheme: ThemeType,
+  iosBuildOptions: {|
+    BUILD_CONFIG_APP_NAME: string
```
[the remark this answers is not in the document]
Comment (maxammann) [reply build-configs/configs/BuildConfigType.js:15]: This seems indeed weird.

File: build-configs/configs/integreat-test-cms/index.js (hunk 2 above)
Comment (klinzo) [reply build-configs/configs/integreat-test-cms/index.js:10]: It should not be different, just forgot to change the `appName` here.

My reason for a separate key is the generation of xcconfig files. If we just map each key (e.g. `appName`, `lightTheme`, ...) to its value, the xcconfig files are probably not valid anymore. Currently therefore only `iosBuildOptions` keys are included in the xcconfig files.

One idea would possible to just include a build config option only in the xcconfig file if the value is of type string. Any other suggestions?

File: fastlane/Fastfile [hunk 6]
```diff
@@ -1,16 +1,9 @@
-require 'yaml'
-
-#
-# `options` contains several parameter:
-#
-# version_code: The new version code
-# version_name: The new version name
-# package_name: The name of the app in Sentry
-# sourcemap_path: The path to the sourcemap file to upload
```
Comment (maxammann) [inline fastlane/Fastfile:9]: Why did you remove this docu? I think we need a way to document options or check for them in the line.

File: fastlane/Fastfile (hunk 3 above)
[the remark this answers is not in the document]
Comment (maxammann) [reply fastlane/Fastfile:77]: I think this is part of another task

File: ios/Integreat.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -284,12 +238,8 @@
 			isa = PBXProject;
 			attributes = {
 				LastUpgradeCheck = 940;
-				ORGANIZATIONNAME = Facebook;
+				ORGANIZATIONNAME = "Tür an Tür - Digitalfabrik";
```
Comment (maxammann) [inline ios/Integreat.xcodeproj/project.pbxproj:241]: lol :joy:

File: package.json [hunk 8]
```diff
@@ -78,6 +78,7 @@
   },
   "devDependencies": {
     "@babel/core": "^7.8.4",
+    "@babel/node": "^7.10.3",
```
Comment (maxammann) [inline package.json:81]: :fire: This upgrade is not good, also check the yarn.lock

**Comment by maxammann** [comment]
I think we can go with having multiple schemes for now. Having the option to build malte is especially helpful right now. This can be removed in the future imo though

File: build-configs/configs/BuildConfigType.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (klinzo) [reply build-configs/configs/BuildConfigType.js:15]: See comment above.

File: fastlane/Fastfile (hunk 6 above)
Comment (klinzo) [reply fastlane/Fastfile:9]: Because it was pretty inconsistent. And if you look at the docu above the function, you may as well look in the function itself to see what options are necessary.

File: package.json (hunk 8 above)
Comment (klinzo) [reply package.json:81]: What upgrade? :thinking: I newly added this

File: fastlane/Fastfile (hunk 6 above)
Comment (klinzo) [reply fastlane/Fastfile:9]: I'll readd it

File: package.json (hunk 8 above)
Comment (maxammann) [reply package.json:81]: Version should be ^7.8.4

File: build-configs/configs/integreat-test-cms/index.js (hunk 2 above)
Comment (klinzo) [reply build-configs/configs/integreat-test-cms/index.js]: Will be flattened and uppercased :)
